source/1_engine/rex_engine/src/string/stringid.cpp

```diff
@@ -0,0 +1,137 @@
+#include "rex_engine/string/stringid.h"
+
+#include "rex_engine/string/stringpool.h"
+#include "rex_std/memory.h"
+#include "rex_std/vector.h"
+
+namespace rex
+{
+  namespace internal
+  {
+    //-------------------------------------------------------------------------
+    StringEntryID make(rsl::string_view characters)
+    {
+      if(characters.empty())
+      {
+        return StringEntryID(); // StringID::is_none() == true
+      }
+
+      return *string_pool::store(characters);
+    }
+  } // namespace internal
+
+  //-------------------------------------------------------------------------
+  /**
+   * Create an empty StringID.
+   */
+  StringID::StringID()
+      : m_comparison_index(0)
+  {
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Create an StringID with characters.
+   */
+  StringID::StringID(rsl::string_view characters)
+      : m_comparison_index(internal::make(characters))
+  {
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Retrieve the hashed value
+   */
+  StringID::operator u32() const
+  {
+    return static_cast<u32>(m_comparison_index);
+  }
+
+  //-------------------------------------------------------------------------
+  bool StringID::operator==(const StringID& other) const
+  {
+    return m_comparison_index == other.m_comparison_index;
+  }
+  //-------------------------------------------------------------------------
+  bool StringID::operator!=(const StringID& other) const
+  {
+    return !(*this == other);
+  }
+
+  //-------------------------------------------------------------------------
+  bool StringID::operator==(const StringEntryID& entryID) const
+  {
+    return m_comparison_index == static_cast<uint32>(entryID);
+  }
+  //-------------------------------------------------------------------------
+  bool StringID::operator!=(const StringEntryID& entryID) const
+  {
+    return m_comparison_index != static_cast<uint32>(entryID);
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Converts an StringID to a readable format
+   */
+  rsl::string_view StringID::to_string_view() const
+  {
+    return string_pool::resolve(m_comparison_index);
+  }
+
+  //-------------------------------------------------------------------------
+  /** True for StringID(), StringID(NAME_None) and StringID("None") */
+  bool StringID::is_none() const
+  {
+    return m_comparison_index == 0;
```

Comment:
can this use an internal invalid id?
it's hashing so it's never be 0 afaik but a proper state that this is an invalid id is probably useful here
I know we're not using SID::None anymore, which I still believe is fine as that shouldn't be exposed to the client.
but internally, I do think it's worth keeping an invalid hash result value making these statements easier to spot when looking at the internal code

Reply:
So what do you propose?

Reply:
I propose that StringID has an internal constexpr static member s_none_state_hash = 0 (or something similarly named)
and all StringIDs initialized with this get initialized with this value.
this also makes it possible to easier debug string ids in the future

Reply:
Yes it is.
![image](https://github.com/New-Creta/Rex/assets/7468362/faa7e4d6-3441-4e17-9766-8602bd9829d5)